coremltools/converters/mil/frontend/_utils.py

```diff
@@ -158,4 +156,95 @@ def get_output_names(outputs):
         output_names = [t.name for t in outputs]
         if all([name is None for name in output_names]):
             output_names = None
-    return output_names
\ No newline at end of file
+    return output_names
+
+
+def multiply(numbers):
+    """
+    :param numbers: list[int]
+    :param return: int
+    """
+    ret = 1
+    for num in numbers:
+        ret *= num
+    return ret
+
+
+def get_perm(src_axes, dst_axes):
+    """
+    :param src_axes: list[int]
+    :param dst_axes: list[int]
+    :return: list[int]
+    """
+    return [src_axes.index(s) for s in dst_axes]
+
+
+def solve_generic_einsum(parsed_vectors, a_var, b_var, name):
+    """
+    :param parsed_vectors: list[list[int]]
+    :param a_var:
+        - var
+        - first input variable
+    :param b_var:
+        - var
+        - second input variable
+    :param name:
+        - str
+        - name tp be assigned to the output var
+
+    :return:
+        - var
+        - output var that contains the einsum result
+    """
+    if any_symbolic(a_var.shape) or any_symbolic(b_var.shape):
+        raise ValueError(
+            "Generic einsum does not support flexible shape."
+        )
+
+    a_axes, b_axes, out_axes = parsed_vectors
+    a_dims = list(a_var.shape)
+    b_dims = list(b_var.shape)
+
+    batched_axes = []
+    reduced_axes = []
+    a_unique_axes = []
+    b_unique_axes = []
+
+    batched_dims = []
+    reduced_dims = []
+    a_unique_dims = []
+    b_unique_dims = []
+
+    for a_dim, a_axis in zip(a_dims, a_axes):
+        if a_axis in b_axes:
+            if a_axis in out_axes:
+                batched_axes.append(a_axis)
+                batched_dims.append(a_dim)
+            else:
+                reduced_axes.append(a_axis)
+                reduced_dims.append(a_dim)
+        else:
+            a_unique_axes.append(a_axis)
+            a_unique_dims.append(a_dim)
+
+    for b_dim, b_axis in zip(b_dims, b_axes):
+        if b_axis not in a_axes:
+            b_unique_axes.append(b_axis)
+            b_unique_dims.append(b_dim)
+
+    a_transposed_axes = batched_axes + a_unique_axes + reduced_axes
+    a = mb.transpose(x=a_var, perm=get_perm(a_axes, a_transposed_axes))
+    a_reshaped_dims = [multiply(batched_dims), multiply(a_unique_dims), multiply(reduced_dims)]
+    a = mb.reshape(x=a, shape=a_reshaped_dims)
```

Comment:
I wonder if we can do a bit more optimization here. 
For instance, identify a few cases to add fewer ops: 
- if transpose and/or reshape is a no op then skip them
- if reshape is a no op and transpose is on the last 2 dims, then use the transpose flag in matmul

Reply:
* The transpose and reshape is going to be eliminated in the no-op elimination graph pass
*  For the second one, I think a more generic way is to add a PYMIL graph pass got capture the `tranpose -> matmat` pattern :) ?

Reply:
That's a good point, I agree.